auth/user_mgt_test.go

```diff
@@ -0,0 +1,716 @@
+		// There are two calls to the server, the first one, on creation retunrs the above []byte
+		// that's how we know the params passed validation
+		// the second call to GetUser, tries to get the user with the returned ID above, it fails
+		// with the following expected error
+		if err.Error() != "cannot find user map[localId:[expectedUserID]]" {
+			t.Error(err)
+		}
+	}
+}
+
+func TestBadUpdateParams(t *testing.T) {
+	badParams := []struct {
+		params         *UserToUpdate
+		expectingError string
+	}{
+		{
+			nil,
+			"params must not be empty for update",
+		}, {
+			&UserToUpdate{},
+			"params must not be empty for update",
+		}, {
+			(&UserToUpdate{}).CustomClaims(map[string]interface{}{"a": strings.Repeat("a", 993)}),
+			fmt.Sprintf(`Custom Claims payload must not exceed %d characters`, maxLenPayloadCC),
+		},
+	}
+
+	for _, res := range reservedClaims {
+		badParams = append(badParams,
+			struct {
+				params         *UserToUpdate
+				expectingError string
+			}{
+				(&UserToUpdate{}).CustomClaims(map[string]interface{}{res: true}),
+				fmt.Sprintf(`claim "%s" is reserved, and must not be set`, res)})
+	}
+
+	for i, test := range badParams {
+		_, err := client.UpdateUser(context.Background(), "outofstruct", test.params)
+		if err == nil {
+			t.Errorf("%d) got no error wanted error %s", i, test.expectingError)
+		}
+		if err.Error() != test.expectingError {
+			t.Errorf(`%d) got error "%s" wanted error "%s"`, i, err.Error(), test.expectingError)
+		}
+	}
+}
+
+func TestUpdateUser(t *testing.T) {
+	s, closer := echoServer([]byte(`{
+		"kind": "identitytoolkit#SetAccountInfoResponse",
+		"localId": "expectedUserID",
+		"email": "tefwfd1234eml5f@test.com",
+		"displayName": "display_name",
+		"passwordHash": "UkVEQUNURUQ=",
+		"providerUserInfo": [
+		 {
+		  "providerId": "password",
+		  "federatedId": "tefwfd1234eml5f@test.com",
+		  "displayName": "display_name"
+		 }
+		],
+		"emailVerified": false
+	   }`), t)
+	defer closer()
+
+	goodParams := []*UserToUpdate{
+		(&UserToUpdate{}).Password("123456"),
+		(&UserToUpdate{}).PhoneNumber("+1"),
+		(&UserToUpdate{}).DisplayName("a"),
+		(&UserToUpdate{}).Email("a@a"),
+		(&UserToUpdate{}).PhoneNumber("+1"),
+		(&UserToUpdate{}).CustomClaims(map[string]interface{}{"a": strings.Repeat("a", 992)}),
+	}
+
+	for _, par := range goodParams {
+		_, err := s.Client.UpdateUser(context.Background(), "expectedUserID", par)
+		// There are two calls to the server, the first one, on creation retunrs the above []byte
+		// that's how we know the params passed validation
+		// the second call to GetUser, tries to get the user with the returned ID above, it fails
+		// with the following expected error
+		if err.Error() != "cannot find user map[localId:[expectedUserID]]" {
+			t.Error(err)
+		}
+	}
+}
+
+func TestBadSetCustomClaims(t *testing.T) {
+	badUserParams := []struct {
+		cc   map[string]interface{}
+		estr string
+	}{{
+		map[string]interface{}{"a": strings.Repeat("a", 993)},
```

Comment:
Also test for something that won't serialize into json (i.e. something for which json.Marshal() will throw an error)